ion-schema/src/constraint.rs

```diff
@@ -731,72 +733,97 @@ impl OrderedElementsConstraint {
-                return Err(Violation::new(
-                    "ordered_elements",
-                    ViolationCode::TypeMismatched,
-                    format!("Expected {occurs_range} of type {type_def}: found {count}"),
-                    ion_path,
-                ));
+    // Builds an NFA state machine based on given type_ids. This is a limited form of NFA where state machine is linear and every transition either leads to itself or the next state.
+    //
+    // All the states has some transitions between them leading from one state to another or back to itself.
+    // All the intermediate states that have a minimum occurrence of 0 are optional states, meaning those states can lead to another state with 0 occurrence event transitions.
+    // There are two special cases of transition that need to be handled.
+    // For any state whose corresponding `type_id` has an `occurs` where:
+    //   * `max >= 2`, that state will have a transition back to itself, allowing for repetition.
+    //   * `min == 0`, that state will have a transition that advances to the next state automatically, making an occurrence of that `type_id` optional.
+    //
+    // Here is an example of how the built NFA would look like for an `ordered_elements` constraint:
+    // ```
+    // ordered_elements: [
+    //     { type: int, occurs: optional },
+    //     number,
+    //     any
+    // ]
+    // ```
+    // NFA:
+    //                              +--------- 0 -----------+
+    //                              |                       |
+    //                              |                       V
+    // I(INITIAL) ----> S1(INTERMEDIATE(0, 1)) -- 1 --> S2(INTERMEDIATE(1, 1)) -- 1 --> S3(INTERMEDIATE(1, 1)) ----> F(FINAL)
+    //
+    // Validation:
+    // Valid input value: `[1, 2, 3]`
+    // +------------------------------+
+    // | event  | State Visits        |
+    // +------------------------------+
+    // |   -    |  I: 1               |
+    // |   1    |  S1: 1, S2: 1       |
+    // |   2    |  S2: 1, S3: 1       |
+    // |   3    |  S3: 1              |
+    // |   END  |  F: 1               |
+    // +------------------------------+
+    //
+    // Invalid input value: `[1, 2]`
+    // +------------------------------+
+    // | event  | State Visits        |
+    // +------------------------------+
+    // |   -    |  I: 1               |
+    // |   1    |  S1: 1, S2: 1       |
+    // |   2    |  S2: 1, S3: 1       |
+    // |   END  |  S3: 1              |
+    // +------------------------------+
+    // As shown above visit count for `END` doesn't have final state in it which means the value resulted to be invalid.
+    //
+    fn build_nfa_from_type_ids(type_ids: &Vec<TypeId>, type_store: &TypeStore) -> NfaRun {
+        let mut nfa_builder = NfaBuilder::new();
+        // add initial state to nfa
+        nfa_builder.with_state(State::Initial);
+
+        for type_id in type_ids {
```

Comment:
Might be useful to have the index of the current element instead of trying to compute it based on the total number of states in the `nfa_builder`.

```suggestion
        for (i, type_id) in type_ids.iter().enumerate() {
```